add air purifier

shossk opened the pull request:
Closes #2

shossk committed:
add purifier

dvcrn commented on src/devices/purifier/class.ts:
```diff
@@ -0,0 +1,217 @@
+	 * Convert mode string to binary.
+	 * @return     {string}  mode binary
+	 */
+	existMode = (mode: string): string => {
```
existMode -> stringToMode or something like this would be better

shossk replied:
All right, thanks!

By the way, would you prefer the private function? I don't plan to use it from outside.

dvcrn replied:
sure, lets make it private

shossk replied:
ok! i'm fix it.

dvcrn commented on src/devices/purifier/class.ts:
```diff
@@ -0,0 +1,217 @@
+	 *
+	 * @param      {PropertyStatus}  propertyStatus  The property status
+	 */
+	queuePropertyStatusUpdate(propertyStatus: PropertyStatus): void {
```
This looks identical to device.ts, so we can remove it here. Use the `super.queuePropertyStatusUpdate`

shossk replied:
There is a reason for this, the type PropertyStatus is slightly different

I didn't understand the extend part of the type, and I thought I would have to change the device.ts property file to make it the same as device.ts, so I temporarily wrote it separately.

I thought that OPERATION_MODE was unique and included only in the air purifier, so I wanted to separate it, is there any way to solve this?

```typescript
export enum StatusCode {
  STATE_DETAIL = 'FA',
  WINDSPEED = 'A0',
  POWER = '80',
  NANOE_MODE = 'CF',
  OPERATION_MODE = 'B0',
  HEAT_TEMPERATURE = 'B3',
  ROOM_TEMPERATURE = 'BB',
}
```

↓

```typescript
export enum StatusCode {
  STATE_DETAIL = 'FA',
  WINDSPEED = 'A0',
  POWER = '80',
  NANOE_MODE = 'CF',
  OPERATION_MODE = 'B0',
  HEAT_TEMPERATURE = 'B3',
  ROOM_TEMPERATURE = 'BB',
  OPERATION_MODE = 'F3', // ← added this
}
```

shossk replied:
I'm sorry, but the solution to this was to make it super and make it as common as possible.
I've made it any, and I know that's not a good solution, so best practices, please!

dvcrn commented on src/devices/purifier/properties.ts:
```diff
@@ -0,0 +1,79 @@
+	RANGE = 'valueRange',
+}
+
+export interface BinaryPropertyStatus {
```
For the types that are identical, just use the ones from the existing properties file - https://github.com/dvcrn/cocoro-sdk/blob/master/src/properties.ts#L72, no need to re-define them

shossk replied:
fix it!

dvcrn commented:
Looking good! I just don't have a way to test it here. I assume you tried all of these settings with your purifier, right?

shossk committed:
Organize property files.
Tweak function to convert string to mode.

shossk commented:
>Looking good! I just don't have a way to test it here. I assume you tried all of these settings with your purifier, right?

Yes! The air purifier in my house is working just fine because of this!

shossk committed:
add private for not use from outside / chunkState

dvcrn commented:
For the public interface of this sdk, we're currently getting devices like this:

```
const devices = await cocoro.queryDevices();
const device = devices[0];
```

Is it doable to automatically deduct when a device is a purifier or aircon?

Maybe we could add a separate type field for each one

```
console.log(device.kind); // => Device.Purifier | Device.Aircon
```

shossk commented:
Do you mean you are looking for a property that can distinguish between an air purifier and an air conditioner?

There is a deviceType as part of the boxinfo, and it looks like you're only looking at part of the boxinfo right now.

I'm sorry if I'm wrong...

```JSON
{
  "box": [
    {
      "echonetData": [
        {
          "labelData": {
            "deviceType": "AIR_CLEANER"
          }
        }
      ]
    }
  ]
}
```

dvcrn commented:
Ah perfect. Okay can you do one last change? Then we can merge this 👍 

In the constructor to device, can you add a new property for`kind`? https://github.com/dvcrn/cocoro-sdk/blob/master/src/cocoro.ts#L122-L134 If deviceType == 'AIR_CLEANER', set it to a new enum type Device.AirCleaner

shossk committed:
Add the process of branching the air purifier

shossk commented:
I've finished adding them!

Added an OR condition for the devices array, remove it if it is not needed.
I'm not sure if it's because I added the OR condition, but the completion doesn't work anymore.
I can see the devices, but the Purifier completion does not work.
Do you have any idea why?

<img src="https://user-images.githubusercontent.com/57487433/150150528-969bae08-2734-42b9-bf60-a8ca86af7696.png" width="250px">

<img src="https://user-images.githubusercontent.com/57487433/150150505-380703c7-2037-45c3-acc3-96737d0d842b.png" width="400px">

dvcrn commented:
Thanks! Looks good enough to me. I'll fix that on my end!

Thanks a lot for taking the time to add support for this! It's been very helpful!

dvcrn commented:
@shossk I've created a separate PR to cleanup some of the code and make things more generic. Can you give it a try to see if your device still works? https://github.com/dvcrn/cocoro-sdk/pull/4

Regarding autocomplete, you need to type-assert the device as Purifier, so (device as Purifier)

shossk commented:
@dvcrn 
Thanks for the cleanups, they work perfectly!
